### PR TITLE
Integrate new device derivatives into `qml.execute`

### PR DESCRIPTION
**Pull request by albi3ro**
This PR integrates device derivatives coming from the experimental device with `qml.execute`.

In doing so, I also updated `DefaultQubit2.preprocess` so that it expands operations that aren't supported by adjoint differentiation if it is requested.

I also added a possible key `"gradient-transform"` to `ExecutionConfig.gradient_method` to correspond to a callable gradient transform. We could consider adding an optional `gradient_transform` key to the `ExecutionConfig` for when the gradient method is a gradient transform.

### CODE REVIEW

**Comment by albi3ro** [comment]
[sc-38764]

File: pennylane/interfaces/execution.py [hunk 1]
```diff
@@ -506,19 +516,54 @@ def cost_fn(params, x):
...
+    elif gradient_fn == "device":
         # gradient function is a device method
 
         # Expand all tapes as per the device's expand function here.
         # We must do this now, prior to the interface, to ensure that
         # decompositions with parameter processing is tracked by the
         # autodiff frameworks.
-        for i, tape in enumerate(tapes):
-            tapes[i] = expand_fn(tape)
+        tapes = tuple(expand_fn(t) for t in tapes)
```
Comment (albi3ro) [inline pennylane/interfaces/execution.py:563]:
```suggestion
        tapes = [expand_fn(t) for t in tapes]
```

File: pennylane/devices/qubit/preprocess.py [hunk 2]
```diff
@@ -95,34 +100,18 @@ def validate_and_expand_adjoint(
...
+        new_ops = [
+            final_op
+            for op in circuit._ops
+            for final_op in _operator_decomposition_gen(op, _accepted_adjoint_operator)
+        ]
+    except RecursionError as e:
+        raise DeviceError(
+            "Reached recursion limit trying to decompose operations. "
+            "Operator decomposition may have entered an infinite loop."
+        ) from e
```
Comment (timmysilv) [inline pennylane/devices/qubit/preprocess.py:114]: is it worth putting this subroutine into its own function, since it's a copy-paste from expand_fn?

File: pennylane/devices/qubit/preprocess.py [hunk 3]
```diff
@@ -95,34 +100,18 @@ def validate_and_expand_adjoint(
         Union[.QuantumTape, .DeviceError]: The expanded tape, such that it is supported by adjoint differentiation.
         If the circuit is invalid for adjoint differentiation, a DeviceError with an explanation is returned instead.
     """
```
Comment (timmysilv) [inline pennylane/devices/qubit/preprocess.py:102]: we should add "Raises: DeviceError" to the docstring. Also, it's starting to get confusing now that it both raises and returns DeviceErrors. Should we always just return it?

File: pennylane/interfaces/execution.py [hunk 4]
```diff
@@ -506,19 +516,54 @@ def cost_fn(params, x):
...
         # decompositions with parameter processing is tracked by the
         # autodiff frameworks.
-        for i, tape in enumerate(tapes):
-            tapes[i] = expand_fn(tape)
+        tapes = [expand_fn(t) for t in tapes]
 
         if gradient_kwargs.get("method", "") == "adjoint_jacobian":
-            mode = "forward" if grad_on_execution else "backward"
-            tapes = _adjoint_jacobian_expansion(tapes, mode, interface, max_expansion)
+            tapes = _adjoint_jacobian_expansion(tapes, grad_on_execution, interface, max_expansion)
```
Comment (timmysilv) [inline pennylane/interfaces/execution.py:566]: what's happened here? were we passing `mode` (a string) to a function that was expecting a bool? how did that ever work?

File: tests/devices/qubit/test_preprocess.py [hunk 5]
```diff
@@ -382,19 +396,22 @@ def test_not_expval(self):
         )
 
     def test_unsupported_op(self):
-        """Test if a QuantumFunctionError is raised for an unsupported operation, i.e.,
-        multi-parameter operations that are not qml.Rot"""
+        """Test that a device supported by the forward pass but"""
```
Comment (timmysilv) [inline tests/devices/qubit/test_preprocess.py:399]: name and docstring need updating. also, why did U2 gain support, deeper decomposition?

**Review by timmysilv — COMMENTED** [review]
overall looks good. I'm a little unsure about adding `gradient-transform` as a supported gradient method. The whole thing makes me think we might want to re-consider our abstraction a bit, so I'd like to take some time on this. It was weird that the old API accepted a Union[str, callable] so maybe we can do better this time around.

File: pennylane/devices/qubit/preprocess.py [hunk 6]
```diff
@@ -95,34 +100,18 @@ def validate_and_expand_adjoint(
...
-            ops = op.decomposition()
-            expanded_ops.extend(ops)
-        elif not isinstance(op, qml.operation.StatePrep):
-            expanded_ops.append(op)
+    try:
+        new_ops = [
+            final_op
+            for op in circuit._ops
+            for final_op in _operator_decomposition_gen(op, _accepted_adjoint_operator)
+        ]
```
Comment (mudit2812) [inline pennylane/devices/qubit/preprocess.py:109]: The reason why previously we only allowed `Rot` to be decomposed and not any other multi-parameter operator is because the decomposition of `Rot` has the same parameters as the ones that are passed in. This is not true for the decompositions of other parametrized operators.

For example, `U2` takes two parameters, so it would get decomposed. However, the operators from the decomposition collectively accept 5 parameters, which would then become the parameters that are differentiated.

It could potentially be fixed with the addition of some type of quantum to classical jacobian post-processing, which is honestly not really worth the trouble in my opinion. I don't even know how we would track the derivatives between the actual parameters and the decomposed parameters in a general way.

Comment (albi3ro) [reply pennylane/devices/qubit/preprocess.py:109]: So the intention of `Device.preprocess` is that its tracked by the machine learning interface. Therefore it can have a classical component to its transformations, see [`execution.py:_adjoint_jacobian_expansion`](https://github.com/PennyLaneAI/pennylane/blob/8af85140f9778e9a171e0e1bf8ebf6cbab42222d/pennylane/interfaces/execution.py#L66). We are basically wanting to get rid of the patch by moving device specific expansion to Device preprocessing.

If we were performing this expansion inside `compute_derivatives` somewhere, we could not decompose `U3`. But since this function is tracked by the machine learning interface, we can have classical components to the derivatives.

Comment (josh146) [reply pennylane/devices/qubit/preprocess.py:109]: > It could potentially be fixed with the addition of some type of quantum to classical jacobian post-processing, which is honestly not really worth the trouble in my opinion. I don't even know how we would track the derivatives between the actual parameters and the decomposed parameters in a general way.

@mudit2812 I'm curious to understand more here 🤔 As long as the processing of the parameters during decomposition is done using AD functions (e.g., `qml.math`), _and_ the decomposition happens before the `custom_vjp`/interface boundary, everything should 'just work'.

**Comment by josh146** [comment]
> It was weird that the old API accepted a Union[str, callable] so maybe we can do better this time around.

I can provide some user-facing context here, not sure if relevant or not(?)

- `diff_method="str"` was the original UI in v0.1 of PL. At this time, gradient transforms were an internal part of the QNode.

- Later on we factored out gradient transforms into their own module, and added an API for end-users to create their own gradient transforms.

- As a result, we wanted to transition to `diff_method=gradient_transform_callable`, to allow users to use their _own_ custom gradient transforms as the diff method, and also as a more scalable approach, as we realized we would have a growing library of gradient transforms over time.

- The plan was to eventually deprecate and remove `diff_method="str"`, but:
  - this was quite user-friendly for beginners,
  - so entrenched/used, and
  - most importantly we had no `callable` equivalent for gradient methods that weren't transforms, e.g., `adjoint` and `device`.
  As a result, we happily left it as `Union[str, Callable]`.

**Comment by albi3ro** [comment]
> overall looks good. I'm a little unsure about adding `gradient-transform` as a supported gradient method. The whole thing makes me think we might want to re-consider our abstraction a bit, so I'd like to take some time on this. It was weird that the old API accepted a Union[str, callable] so maybe we can do better this time around.

The Execution config is probably under a bit too much flux right now. We should probably try to lock it down a bit mroe.

I 100% want a separation between string descriptions and callable descriptions.  I'm also now remembering that AWS wanted to be able to make "parameter-shift" a "device derivative".  That way the remote devices can handle parameter shift remotely so they don't have to send so many tapes back and forth.

I'm also realizing we probably shouldn't limit the possible gradient methods in the execution config.  We should leave it open for the device to have their own custom type of derivative.  As long as the device says "I support this" by a property like `use_device_derivative`, pennylane shouldn't care.

**Comment by albi3ro** [comment]
> > It was weird that the old API accepted a Union[str, callable] so maybe we can do better this time around.
> 
> I can provide some user-facing context here, not sure if relevant or not(?)
> 
> * `diff_method="str"` was the original UI in v0.1 of PL. At this time, gradient transforms were an internal part of the QNode.
> * Later on we factored out gradient transforms into their own module, and added an API for end-users to create their own gradient transforms.
> * As a result, we wanted to transition to `diff_method=gradient_transform_callable`, to allow users to use their _own_ custom gradient transforms as the diff method, and also as a more scalable approach, as we realized we would have a growing library of gradient transforms over time.
> * The plan was to eventually deprecate and remove `diff_method="str"`, but:
>   
>   * this was quite user-friendly for beginners,
>   * so entrenched/used, and
>   * most importantly we had no `callable` equivalent for gradient methods that weren't transforms, e.g., `adjoint` and `device`.
>   
>   As a result, we happily left it as `Union[str, Callable]`.

I think we can keep the overloading at the top-level, ie construction of the QNode, but I think we can probably start to separate them out internally, especially if we are storing the information in a configuration object instead of a space limited list of keyword arguments.

File: pennylane/devices/qubit/preprocess.py (hunk 6 above)
Comment (mudit2812) [reply pennylane/devices/qubit/preprocess.py:109]: Thanks for the context! There's a lot about derivative computation in Pennylane that I'm very unfamiliar with. I have more clarity on it now :)

File: pennylane/interfaces/execution.py [hunk 7]
```diff
@@ -21,7 +21,7 @@
 
 # pylint: disable=import-outside-toplevel,too-many-arguments,too-many-branches,not-callable
 # pylint: disable=unused-argument,unnecessary-lambda-assignment,inconsistent-return-statements,
-# pylint: disable=too-many-statements, invalid-unary-operand-type
+# pylint: disable=too-many-statements, invalid-unary-operand-type, function-redefined
```
Comment (mudit2812) [inline pennylane/interfaces/execution.py:24]: The amount of pylint warnings that need to be disabled for this file to work is making me sweat.

Comment (albi3ro) [reply pennylane/interfaces/execution.py:24]: And that's not even getting into the number of sourcery complaints I get locally.

**Review by mudit2812 — APPROVED** [review]

File: tests/devices/qubit/test_preprocess.py (hunk 5 above)
Comment (albi3ro) [reply tests/devices/qubit/test_preprocess.py:399]: This PR also updates the preprocessing to perform deeper decompositions  (till no variables or one variable with a generator) in the situation where adjoint differentiation was requested.

**Review by timmysilv — APPROVED** [review]
my concern remains about the config, but I won't block this PR for it. Also, I'll copy your autograd test changes to the other interface tests when I open the Torch PR

File: pennylane/interfaces/jax.py [hunk 8]
```diff
@@ -488,7 +488,7 @@ def execute_wrapper_jvp(primals, tangents):
                         max_diff=max_diff,
                     )
                 )
-
+            print(primals[0])
```
Comment (albi3ro) [inline pennylane/interfaces/jax.py:491]:
```suggestion

```

File: pennylane/devices/qubit/preprocess.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/devices/qubit/preprocess.py:114]: Probably. We can probably come back through and tidy all this up at some point.

File: pennylane/devices/qubit/preprocess.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/devices/qubit/preprocess.py:102]: Given that the recursion error is enough an extreme edge case, I'm not too concerned.
